File: data_structures/3_python/c_stacks.py
class CustomStack:
    def __init__(self):
        self.elements = []

    # MARK: - Stack Operations

    # Check if the stack is empty.
    def is_empty(self):
        return len(self.elements) == 0

    # Get the number of elements in the stack.
    def count(self):
        return len(self.elements)

    # Push an element onto the stack.
    def push(self, element):
        self.elements.append(element)

    # Pop the top element from the stack.
    def pop(self):
        if self.elements:
            return self.elements.pop()
        else:
            return None

    # Peek at the top element in the stack without removing it.
    def peek(self):
        if self.elements:
            return self.elements[-1]
        else:
            return None

    # MARK: - Stack Operations with List

    # Initialize the stack with a list of elements.
    def initialize_with_list(self, lst):
        self.elements = lst

    # Push a list of elements onto the stack.
    def push_list(self, lst):
        self.elements.extend(lst)

    # Pop a specified number of elements from the stack.
    def pop_list(self, count):
        popped_elements = []
        for _ in range(count):
            if self.elements:
                popped_elements.append(self.elements.pop())
            else:
                break
        return popped_elements

    # MARK: - Clear and Reset

    # Remove all elements from the stack.
    def clear(self):
        self.elements = []

    # MARK: - Checking for Element

    # Check if the stack contains a specific element.
    def contains(self, element):
        return element in self.elements

    # MARK: - Filtering

    # Filter the stack using a given predicate.
    def filter(self, predicate):
        return list(filter(predicate, self.elements))

    # MARK: - Conversion to List

    # Convert the stack to a list.
    def to_list(self):
        return self.elements

    # MARK: - Map

    # Transform each element in the stack using a provided transform function.
    def map(self, transform):
        return [transform(element) for element in self.elements]

    # MARK: - Reduce

    # Combine all elements in the stack using a reducer function.
    def reduce(self, initial_result, reducer):
        result = initial_result
        for element in self.elements:
            result = reducer(result, element)
        return result

    # MARK: - Concatenate

    # Concatenate another stack to this stack.
    def concatenate(self, other_stack):
        self.elements.extend(other_stack.elements)

    # MARK: - Subscript

    # Access the element at a specific index in the stack.
    def __getitem__(self, index):
        if 0 <= index < len(self.elements):
            return self.elements[index]
        else:
            return None

    # ... Add more stack operations as needed ...


# Test the CustomStack class
if __name__ == "__main__":
    stack = CustomStack()
    stack.push(1)
    stack.push(2)
    stack.push(3)

    print("Stack:")
    for i in range(stack.count() - 1, -1, -1):
        element = stack[i]
        if element is not None:
            print(element)

File: data_structures/5_kotlin/c_stacks.kt
class CustomStack<T> {
    private val elements: MutableList<T> = mutableListOf()

    // MARK: - Stack Operations

    /// Check if the stack is empty.
    fun isEmpty(): Boolean {
        return elements.isEmpty()
    }

    /// Get the number of elements in the stack.
    fun count(): Int {
        return elements.size
    }

    /// Push an element onto the stack.
    fun push(element: T) {
        elements.add(element)
    }

    /// Pop the top element from the stack.
    fun pop(): T? {
        return if (elements.isNotEmpty()) elements.removeAt(elements.size - 1) else null
    }

    /// Peek at the top element in the stack without removing it.
    fun peek(): T? {
        return elements.lastOrNull()
    }

    // MARK: - Stack Operations with List

    /// Initialize the stack with a list of elements.
    fun initializeWithList(list: List<T>) {
        elements.clear()
        elements.addAll(list)
    }

    /// Push a list of elements onto the stack.
    fun pushList(list: List<T>) {
        elements.addAll(list)
    }

    /// Pop a list of elements from the stack.
    fun popList(count: Int): List<T> {
        val poppedElements = mutableListOf<T>()
        repeat(count) {
            if (elements.isNotEmpty()) {
                poppedElements.add(elements.removeAt(elements.size - 1))
            }
        }
        return poppedElements
    }

    // MARK: - Clear and Reset

    /// Remove all elements from the stack.
    fun clear() {
        elements.clear()
    }

    // MARK: - Checking for Element

    /// Check if the stack contains a specific element.
    fun contains(element: T): Boolean {
        return elements.contains(element)
    }

    // MARK: - Filtering

    /// Filter the stack using a given predicate.
    fun filter(predicate: (T) -> Boolean): List<T> {
        return elements.filter(predicate)
    }

    // MARK: - Conversion to List

    /// Convert the stack to a list.
    fun toList(): List<T> {
        return elements.toList()
    }

    // MARK: - Map

    /// Transform each element in the stack using a provided transform function.
    fun <U> map(transform: (T) -> U): List<U> {
        return elements.map(transform)
    }

    // MARK: - Reduce

    /// Combine all elements in the stack using a reducer function.
    fun <Result> reduce(initialResult: Result, operation: (Result, T) -> Result): Result {
        return elements.fold(initialResult, operation)
    }

    // MARK: - Concatenate

    /// Concatenate another stack to this stack.
    fun concatenate(otherStack: CustomStack<T>) {
        elements.addAll(otherStack.elements)
    }

    // MARK: - Subscript

    /// Access the element at a specific index in the stack.
    operator fun get(index: Int): T? {
        return if (index in 0 until elements.size) elements[index] else null
    }

    // ... Add more stack operations as needed ...
}

fun main() {
    val stack = CustomStack<Int>()
    stack.push(1)
    stack.push(2)
    stack.push(3)

    println("Stack:")
    stack.toList().forEach { println(it) }
}
